br/pkg/restore/snap_client/pipeline_items.go

```diff
@@ -118,154 +116,262 @@ func (rc *SnapClient) RestorePipeline(ctx context.Context, plCtx PipelineContext
+func (handler *PipelineConcurrentHandler) afterTableRestoredCh(createdTables []*CreatedTable) <-chan *CreatedTable {
 	outCh := make(chan *CreatedTable)
 
-	go func() {
+	handler.eg.Go(func() error {
 		defer close(outCh)
 
 		for _, createdTable := range createdTables {
 			select {
-			case <-ctx.Done():
-				return
-			default:
-				outCh <- createdTable
+			case <-handler.pipelineTaskCtx.Done():
+				return handler.pipelineTaskCtx.Err()
+			case outCh <- createdTable:
 			}
 		}
-	}()
+
+		return nil
+	})
 	return outCh
 }
 
 // dropToBlackhole drop all incoming tables into black hole,
 // i.e. don't execute checksum, just increase the process anyhow.
-func dropToBlackhole(ctx context.Context, inCh <-chan *CreatedTable, errCh chan<- error) <-chan struct{} {
-	outCh := make(chan struct{}, 1)
-	go func() {
-		defer func() {
-			close(outCh)
-		}()
+func (handler *PipelineConcurrentHandler) dropToBlackhole(inCh <-chan *CreatedTable) {
+	handler.eg.Go(func() error {
 		for {
 			select {
-			case <-ctx.Done():
-				errCh <- ctx.Err()
-				return
+			case <-handler.pipelineTaskCtx.Done():
+				return handler.pipelineTaskCtx.Err()
 			case _, ok := <-inCh:
 				if !ok {
-					return
+					return nil
 				}
 			}
 		}
-	}()
-	return outCh
+	})
 }
 
-func concurrentHandleTablesCh(
-	ctx context.Context,
+func (handler *PipelineConcurrentHandler) concurrentHandleTablesCh(
 	inCh <-chan *CreatedTable,
-	outCh chan<- *CreatedTable,
-	errCh chan<- error,
-	workers *tidbutil.WorkerPool,
+	concurrency uint,
+	taskLabel string,
 	processFun func(context.Context, *CreatedTable) error,
-	deferFun func()) {
-	eg, ectx := errgroup.WithContext(ctx)
-	defer func() {
-		if err := eg.Wait(); err != nil {
-			errCh <- err
-		}
-		close(outCh)
-		deferFun()
-	}()
-
-	for {
-		select {
-		// if we use ectx here, maybe canceled will mask real error.
-		case <-ctx.Done():
-			errCh <- ctx.Err()
-		case tbl, ok := <-inCh:
-			if !ok {
+	deferFun func(context.Context) error,
+) (outCh chan *CreatedTable) {
+	outCh = defaultOutputTableChan()
+	handler.eg.Go(func() (pipelineErr error) {
+		workers := tidbutil.NewWorkerPool(concurrency, taskLabel)
+		eg, ectx := errgroup.WithContext(handler.pipelineTaskCtx)
+		defer func() {
+			// Note: directly return the error and then the pipelineTaskCtx will be cancelled.
+			if err := eg.Wait(); err != nil {
+				log.Error("pipeline item execution is failed", zap.String("task", taskLabel), zap.Error(err))
+				pipelineErr = errors.Trace(err)
 				return
 			}
-			cloneTable := tbl
-			worker := workers.ApplyWorker()
-			eg.Go(func() error {
-				defer workers.RecycleWorker(worker)
-				err := processFun(ectx, cloneTable)
-				if err != nil {
-					return err
+			if err := deferFun(handler.pipelineTaskCtx); err != nil {
+				log.Error("pipeline defer execution is failed", zap.String("task", taskLabel), zap.Error(err))
+				pipelineErr = errors.Trace(err)
+				return
```

Comment:
if error occurs. is it safe to skip close(outCh)?

Reply:
The `handler.pipelineTaskCtx` will be cancelled because the `pipelineErr` is returned in the function of `eg.Go`.
Therefore, the next pipeline(also all the pipelines) will be cancelled by `<-handler.pipelineTaskCtx.Done()`.

Reply:
We need to ensure, there is always select `handler.pipelineTaskCtx.Done()` when channel `send`/`receive` in each pipeline.